model/DataObjectSchema.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+use SilverStripe\Framework\Core\Configurable;
+use SilverStripe\Framework\Core\Injectable;
+use SilverStripe\Model\FieldType\DBComposite;
+
+/**
+ * Provides dataobject and database schema mapping functionality
+ */
+class DataObjectSchema {
+	use Injectable;
+	use Configurable;
+
+	/**
+	 * Default separate for table namespaces. Can be set to any string for
+	 * databases that do not support some characters.
+	 *
+	 * Defaults to \ to to conform to 3.x convention.
+	 *
+	 * @config
+	 * @var string
+	 */
+	private static $table_namespace_separator = '\\';
+
+	/**
+	 * Cache of database fields
+	 *
+	 * @var array
+	 */
+	protected $databaseFields = [];
+
+	/**
+	 * Cache of composite database field
+	 *
+	 * @var array
+	 */
+	protected $compositeFields = [];
+
+	/**
+	 * Cache of table names
+	 *
+	 * @var array
+	 */
+	protected $tableNames = [];
+
+	/**
+	 * Clear cached table names
+	 */
+	public function reset() {
+		$this->tableNames = [];
+		$this->databaseFields = [];
+		$this->compositeFields = [];
+	}
+
+	/**
+	 * Get all table names
+	 *
+	 * @return array
+	 */
+	public function getTableNames() {
+		$this->cacheTableNames();
+		return $this->tableNames;
+	}
+
+	/**
+	 * Given a column name, build the "Table"."Column" string appropriate for querying with.
+	 * This method will correctly match the column to the data class and table.
+	 *
+	 * Note that any fixed_fields will use the base table. Use quotedID() for joining
+	 * on the ID of non-base tables.
+	 *
+	 * @param string $class Class name
+	 * @param string $column Column name
+	 * @return string The quoted column, if this column belongs to this model
+	 */
+	public function quotedColumn($class, $column) {
```

Comment:
In principle, fields could be remapped as well as tables. To better support this, I think that the 2nd argument of `quotedColumn()` should be `$field`, not `$column`.

 * The inputs, class and field, are names used with the domain model
 * The output, a quoted table.column pair, refers to names used in the database persistence layer 

It doesn't really affect much other than the docblock, but it'll make things clearer for future developers.